### PR TITLE
Resolve circular reference in ThrottledFunc

### LINKED ISSUE #7710: Terminal gets extremely sluggish after a day of use
I'm finding that if I use Windows Terminal for more than a day, it starts to get extremely sluggish: there is a lag in opening and closing new tabs and a lag in switching between tabs and a lag in typing. This problem immediately resolves itself if I close the application and restart it.

I'm not sure how to give a more detailed bug report. I tried downloading WPR, but it errored out when I tried to make a recording.

My setup is as follows: I'm using the beta version of Terminal, running WSL2 with ubuntu 20.04. I'm on the insider branch of Windows 10 with CUDA support for WSL2 enabled. My laptop is a Razer Blade advanced 2019 4K with an nvidia 2080 mobile graphics card.

### PR DESCRIPTION
**Pull request by lhecker**
## Summary of the Pull Request

ThrottledFunc previously created a DispatcherTimer whose Tick callback holds a strong reference to the DispatcherTimer itself.
This causes a reference cycle, inadvertently leaking timer instances.

## PR Checklist

* [x] Closes #7710
* [x] I work here

## Detailed Description of the Pull Request / Additional comments

I've initially wanted to remove the `ThrottledFunc<>` optimization, but it turns out that this causes a 3% slowdown. That's definitely not a lot, but enough that we can just keep the optimization for the time being.
I've moved the implementation from the .cpp file into the header regardless since the two implementations are extremely similar and it's easier that way to keep them in line.

## Validation Steps Performed

I've ensured that the scrollbar still updates its length when I add new lines to a newly created tab.

### CODE REVIEW

**Comment by miniksa** [comment]
> ## References
> This PR potentially solves the issues seen in #7710.

Pretty sure it has to say "Closes <issue number>" for GitHub to automatically pick up the connection and tie them together.

**Comment by lhecker** [comment]
@miniksa I wasn't sure whether I should write that it closes that issue since we actually don't know what the cause for the severe slowdown is for the affected people. While we both can reproduce the timer leak locally, neither of us could reproduce a slowdown that causes such severe issues though, right?
But you gave me the idea to try and replicate it by _slowly_ dumping a lot of text in a lot of tabs. I think the "slow" part is what amplifies the issue.

**Comment by lhecker** [comment]
I was able to replicate the issue by running the following "slow cat" in 30 tabs simultaneously:
```fish
cat big.txt | while read l; echo $l; sleep 0.05; end
```
After about 10min the current store version started feeling sluggish, whereas this one didn't.
--> This PR should fix #7710 entirely.

File: src/cascadia/TerminalControl/ThrottledFunc.h [hunk 1]
```diff
@@ -110,12 +90,29 @@ class ThrottledFunc : public std::enable_shared_from_this<ThrottledFunc<Args...>
     }
 
 private:
+    static winrt::fire_and_forget _Fire(winrt::Windows::Foundation::TimeSpan delay, winrt::Windows::UI::Core::CoreDispatcher dispatcher, std::weak_ptr<ThrottledFunc> weakThis)
```
[the remark this answers is not in the document]
Comment (miniksa) [reply src/cascadia/TerminalControl/ThrottledFunc.h:93]: Sounds great to me. Good point on avoiding accidental usage.

**Review by miniksa — COMMENTED** [review]
Let me pull this branch and do the heap trace to see if the leak is gone.

**Comment by miniksa** [comment]
> @miniksa I wasn't sure whether I should write that it closes that issue since we actually don't know what the cause for the severe slowdown is for the affected people. While we both can reproduce the timer leak locally, neither of us could reproduce a slowdown that causes such severe issues though, right?
> But you gave me the idea to try and replicate it by _slowly_ dumping a lot of text in a lot of tabs. I think the "slow" part is what amplifies the issue.

You're right that we're not 100% sure this will fix the issue for folks until they try it out, but we have a lot of reasons to be confident in it. It's a slow leak over time. It's on the UI dispatcher thread which makes everything slow. This throttler was introduced in June and the bug opened in September so by time proximity and the release schedule, it correlates. And the root cause and resolution make sense for the experience people are having.

Also... we can always reopen the issue or duplicate it into a new one if we close it and it turns out to not work. So yeah. Just go for "Closes" when you're relatively confident.

**Review by miniksa — APPROVED** [review]

File: src/cascadia/TerminalControl/ThrottledFunc.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (DHowett) [reply src/cascadia/TerminalControl/ThrottledFunc.h:93]: So cool that we got to move it to the header!

**Review by DHowett — APPROVED** [review]
EX CELLENT

**Comment by DHowett** [comment]
@msftbot merge this in like 2ish minutes

**Comment by DHowett** [comment]
@msftbot merge this in 1 minute

**Pull request merged**